pkg/agent/pv_monitor_agent.go

```diff
@@ -0,0 +1,201 @@
+	pvListerSynced cache.InformerSynced
+
+	pvcLister       corelisters.PersistentVolumeClaimLister
+	pvcListerSynced cache.InformerSynced
+
+	podLister       corelisters.PodLister
+	podListerSynced cache.InformerSynced
+
+	supportStageUnstage bool
+	kubeletRootPath     string
+	pvChecker           *handler.PVHealthConditionChecker
+
+	podWithPVItemQueue workqueue.Interface
+}
+
+// NewPVMonitorAgent create pv monitor agent
+func NewPVMonitorAgent(client kubernetes.Interface, monitorName string, conn *grpc.ClientConn, timeout time.Duration, monitorInterval time.Duration, pvInformer coreinformers.PersistentVolumeInformer,
+	pvcInformer coreinformers.PersistentVolumeClaimInformer, podInformer coreinformers.PodInformer, supportStageUnstage bool, kubeletRootPath string) *PVMonitorAgent {
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: client.CoreV1().Events(v1.NamespaceAll)})
+	var eventRecorder record.EventRecorder
+	eventRecorder = broadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: fmt.Sprintf("csi-pv-monitor-controller-%s", monitorName)})
+
+	agent := &PVMonitorAgent{
+		supportStageUnstage: supportStageUnstage,
+		kubeletRootPath:     kubeletRootPath,
+		csiConn:             conn,
+		eventRecorder:       eventRecorder,
+		client:              client,
+		monitorName:         monitorName,
+		monitorInterval:     monitorInterval,
+		podWithPVItemQueue:  workqueue.NewNamed("csi-monitor-pod-pv-queue"),
+	}
+
+	// PV lister
+	agent.pvLister = pvInformer.Lister()
+	agent.pvListerSynced = pvInformer.Informer().HasSynced
+
+	// PVC informer
+	agent.pvcLister = pvcInformer.Lister()
+	agent.pvListerSynced = pvcInformer.Informer().HasSynced
+
+	// Pod informer
+	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    agent.podAdded,
+		UpdateFunc: agent.podUpdated,
+		// deleted pods will not be readded to the queue, do not need DeleteFunc here
+	})
+	agent.podLister = podInformer.Lister()
+	agent.podListerSynced = podInformer.Informer().HasSynced
+
+	agent.pvChecker = handler.NewPVHealthConditionChecker(monitorName, conn, client, timeout, agent.pvcLister, agent.pvLister, agent.eventRecorder)
+
+	return agent
+}
+
+// Run runs volume health condition checking method
+func (agent *PVMonitorAgent) Run(workers int, stopCh <-chan struct{}) {
+	defer agent.podWithPVItemQueue.ShutDown()
+
+	klog.Infof("Starting CSI External PV Health Monitor Agent")
+	defer klog.Infof("Shutting down CSI External PV Health Monitor Agent")
+
+	if !cache.WaitForCacheSync(stopCh, agent.pvcListerSynced, agent.pvListerSynced, agent.podListerSynced) {
+		klog.Errorf("Cannot sync cache")
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(agent.checkPVWorker, agent.monitorInterval, stopCh)
+	}
+
+	// TODO: reconcile Pods and PVs
+
+	<-stopCh
+}
+
+func (agent *PVMonitorAgent) atLeastOneFieldIsEmpty(podWithPVItem *PodWithPVItem) bool {
```

Comment:
This is a weird function name.  Can you rename to something such as IsPodWithPVItemValid?